runbook/executor.go

```diff
@@ -26,7 +30,23 @@ type ExecError struct {
 	error
 }
 
-func Execute(executablePath string, args, environmentVars []string, stdout, stderr io.Writer) error {
+func Execute(messageId string, executablePath string, args, environmentVars []string, stdout, stderr io.Writer) (string, error) {
+
+	callbackContextBuffer := make([]byte, 4096)
+
+	var waitGroup sync.WaitGroup
+
+	var pipePath string
+	if runtime.GOOS == "windows" {
+		pipePath = `\\.\pipe\jecCallbackPipe-` + messageId
+	} else {
+		pipePath = "jecCallbackPipe-" + messageId
+	}
+
+	pipe := CreatePipeFunc(pipePath)
+
+	waitGroup.Add(1)
+	go CallbackContextReaderFunc(callbackContextBuffer, pipe)
```

Comment:
not sure if we need a separate go routine here. is it a significant performance improver?

Reply:
Actually its not for performance improvement. when we open the named pipe for reading (or for windows when we call listener.Accept()) the process will stop there until any other process opens the other end of pipe for writing the data. 
So with this go routine we will parallelly start waiting for pipe write and the main routine will continue to execute the script. 
Before the script execution started we have already created the pipe and waiting for data to be written. Now inside script execution we can open the pipe and write data to it. 

If the script has written some data then the go routine will terminate after reading it. But if the script has not opened the pipe and finished execution then we can call **waitGroup.Done()** to stop that go rountine.

Reply:
do we need to listen to the pipe in parallel? do we have an option like open file -> execute -> read all -> close conn?